components/tikv_util/src/memory.rs

```diff
@@ -65,3 +74,105 @@ impl HeapSize for RaftCmdRequest {
             + mem::size_of_val(&self.status_request)
     }
 }
+
+pub struct MemoryQuota {
+    capacity: AtomicUsize,
+    in_use: AtomicUsize,
+}
+
+impl MemoryQuota {
+    pub fn new(capacity: usize) -> MemoryQuota {
+        MemoryQuota {
+            capacity: AtomicUsize::new(capacity),
+            in_use: AtomicUsize::new(0),
+        }
+    }
+
+    pub fn in_use(&self) -> usize {
+        self.in_use.load(Ordering::Relaxed)
+    }
+
+    pub fn capacity(&self) -> usize {
+        self.capacity.load(Ordering::Acquire)
+    }
+
+    pub fn set_capacity(&self, capacity: usize) {
+        self.capacity.store(capacity, Ordering::Release)
+    }
+
+    pub fn alloc(&self, bytes: usize) -> bool {
```

Comment:
I think this can be further improved by allocating a chunk locally (consumed by alloc) to avoid use compare_exchange_weak for each call. Otherwise, this can be a considerable bottleneck on NUMA architecture, if we hope to use this structure for general memory control such as read.

Reply:
👍🏾  I will address the comment in a later PR.